applications/FluidDynamicsApplication/custom_elements/embedded_fluid_element.h

```diff
@@ -0,0 +1,389 @@
+
+    void DefineCutGeometryData(EmbeddedElementData& rData) const;
+
+    void NormalizeInterfaceNormals(typename EmbeddedElementData::InterfaceNormalsType& rNormals, double Tolerance) const;
+
+    /**
+    * This functions adds the penalty extra term level set contribution.
+    * @param rLHS: reference to the LHS matrix
+    * @param rRHS: reference to the RHS vector
+    * @param rData: reference to element data structure
+    */
+    void AddBoundaryConditionPenaltyContribution(
+        MatrixType& rLHS,
+        VectorType& rRHS,
+        const EmbeddedElementData& rData) const;
+    
+    /**
+     * This function computes the penalty coefficient for the level set BC imposition
+     * @param rLeftHandSideMatrix: reference to the LHS matrix
+     * @param rData: reference to element data structure
+     */
+    double ComputePenaltyCoefficient(const EmbeddedElementData& rData) const;
+
+    /**
+    * This drops the outer nodes velocity constributions in both LHS and RHS matrices.
+    * @param rLHS: reference to the LHS matrix
+    * @param rRHS: reference to the RHS vector
+    * @param rData: reference to element data structure
+    */
+    void DropOuterNodesVelocityContribution(
+        MatrixType& rLHS,
+        VectorType& rRHS,
+        const EmbeddedElementData& rData) const;
+
+    /**
+    * This functions adds the level set strong boundary condition imposition contribution.
+    * @param rLHS: reference to the LHS matrix
+    * @param rRHS: reference to the RHS vector
+    * @param rData: reference to element data structure
+    */
+    void AddBoundaryConditionModifiedNitscheContribution(
+        MatrixType& rLHS,
+        VectorType& rRHS,
+        const EmbeddedElementData& rData) const;
+
+    ///@}
+    ///@name Protected Operations
+    ///@{
+
+
+    ///@}
+    ///@name Protected  Access
+    ///@{
+
+
+    ///@}
+    ///@name Protected Inquiry
+    ///@{
+
+
+    ///@}
+    ///@name Protected LifeCycle
+    ///@{
+
+    ///@}
+
+private:
+
+    ///@name Static Member Variables
+    ///@{
+
+    ///@}
+    ///@name Member Variables
+    ///@{
+
+    ///@}
+    ///@name Serialization
+    ///@{
+
+    friend class Serializer;
+
+    void save(Serializer& rSerializer) const override;
+
+    void load(Serializer& rSerializer) override;
+
+    ///@}
+    ///@name Private Operators
+    ///@{
+
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+
+    ///@}
+    ///@name Private  Access
+    ///@{
+
+
+    ///@}
+    ///@name Private Inquiry
+    ///@{
+
+
+    ///@}
+    ///@name Un accessible methods
+    ///@{
+
+    /// Assignment operator.
+    EmbeddedFluidElement& operator=(EmbeddedFluidElement const& rOther);
+
+    /// Copy constructor.
+    EmbeddedFluidElement(EmbeddedFluidElement const& rOther);
+
+    ///@}
+
+
+}; // Class EmbeddedFluidElement
+
+namespace Internals {
+
+template <size_t TDim, size_t TNumNodes>
+ModifiedShapeFunctions::Pointer GetShapeFunctionCalculator(
```

Comment:
This is not a matrix, it is a complete class by @rubenzorrilla that manages split elements. I use a (base class) pointer because different derived classes will be used by different embedded implementations.